test/distributed/_composable/fsdp/test_fully_shard_memory.py

```diff
@@ -0,0 +1,157 @@
+            + sum(p.numel() for p in model.pos_embeddings.parameters())
+            + sum(p.numel() for p in model.norm.parameters())
+            + sum(p.numel() for p in model.output.parameters())
+        )
+        fully_shard_fn = functools.partial(
+            fully_shard, reshard_after_forward=reshard_after_forward
+        )
+        for module in model.modules():
+            if isinstance(module, TransformerBlock):
+                fully_shard_fn(module)
+        fully_shard_fn(model)
+        # Do not use foreach since intermediates increase peak memory
+        optim = torch.optim.Adam(model.parameters(), lr=1e-2, foreach=False)
+
+        # Init: Each module is moved to GPU before sharding parameters
+        peak_mem_mb = self._get_peak_active_memory_mb()
+        curr_mem_mb = self._get_curr_active_memory_mb()
+        init_mem_mb = (model_sharded_numel + max_unsharded_numel) * 4 / 1e6
+        # Allow for some buffer for the peak memory since original parameters
+        # are not freed until a `fully_shard` call returns
+        buffer_mb = 4
+        self.assertLessEqual(peak_mem_mb - base_mem_mb, init_mem_mb + buffer_mb)
+        self.assertLessEqual(curr_mem_mb - base_mem_mb, init_mem_mb)
+
+        # Use a small input to minimize activation memory usage
+        inp = torch.randint(0, vocab_size, (1, 4), device="cuda")
+
+        # Forward:
+        loss = model(inp)
+        mem_mb = self._get_peak_active_memory_mb()
+        # Allow for some buffer for fragmentation/activations (where this
+        # number is kept much smaller than the actual memory usage, which is on
+        # the order of 100-200+ MB)
+        buffer_mb = 16
+        if reshard_after_forward:
+            # 3x max unsharded block parameters (current all-gather + copy-out
+            # and next all-gather), non-block parameters, and other
+            expected_mem_mb = (
+                3 * max_unsharded_numel + non_block_numel
+            ) * 4 / 1e6 + buffer_mb
+            # Sharded parameters
+            expected_mem_mb += model_sharded_numel * 4 / 1e6
+        else:
+            # Sharded parameters, unsharded parameters, 1x max unsharded block
+            # parameters (copy-out) and other (peak at end of forward)
+            expected_mem_mb = (
+                model_sharded_numel + model_unsharded_numel + max_unsharded_numel
+            ) * 4 / 1e6 + buffer_mb
+        self.assertLessEqual(mem_mb - base_mem_mb, expected_mem_mb)
+
+        # Backward:
+        loss.sum().backward()
+        mem_mb = self._get_peak_active_memory_mb()
+        if reshard_after_forward:
+            # 1x max unsharded block parameters (all-gather), 2.5x max
+            # unsharded block gradients (gradients, reduce-scatter input,
+            # reduce-scatter output), non-block parameters, and other
+            expected_mem_mb = (
+                3.5 * max_unsharded_numel + non_block_numel
```

Comment:
Changed locally -- will push next time:
```suggestion
                3 * max_unsharded_numel + non_block_numel
```
Since we already include `model_sharded_numel * 4` below as part of the "2x sharded parameters/gradients", we do not need to count it as the 0.5 factor here.